File: src-tauri/src/k8s/cost/pricing.rs
//! Pricing resolver backed by the daily-refreshed dataset published as
//! GitHub release assets (see `.github/workflows/update-pricing.yml`).
//!
//! Strategy:
//! 1. For each unique provider in the requested nodes, ensure the provider's
//!    dataset is loaded (memory → disk cache → network fetch, in that order).
//! 2. Resolve `(provider, region, instance_type) -> $/hour` purely in-memory.
//!
//! Cache layers:
//! - In-memory `HashMap<provider, ProviderDataset>` with 24h TTL — survives the
//!   process lifetime, refreshed lazily.
//! - Disk cache at `~/.kdashboard/pricing/<provider>.json` with the same TTL —
//!   survives restarts and offline boots.
//!
//! Failure mode: if both network and disk fail, returns `None` and the caller
//! falls back to hardcoded rates. If the network fails but a stale disk cache
//! exists, we use the stale cache rather than nothing.

use anyhow::{Context, Result};
use serde::Deserialize;
use std::collections::HashMap;
use std::path::PathBuf;
use std::time::{Duration, Instant, SystemTime};
use tokio::sync::Mutex;

use super::types::NodeInfo;

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const PRICING_BASE_URL: &str =
    "https://github.com/folio-pro/kdashboard/releases/download/pricing-data";

const SUPPORTED_PROVIDERS: &[&str] = &["aws", "azure", "gcp"];

const DATASET_TTL: Duration = Duration::from_secs(86400); // 24 hours

const DISK_CACHE_DIR: &str = ".kdashboard/pricing";

// ---------------------------------------------------------------------------
// Dataset shape (matches scripts/fetch-pricing.ts output)
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
struct InstancePrice {
    #[allow(dead_code)]
    vcpu: f64,
    #[allow(dead_code)]
    memory_gb: f64,
    ondemand_usd_hour: Option<f64>,
    #[allow(dead_code)]
    spot_usd_hour: Option<f64>,
}

#[derive(Debug, Deserialize)]
struct RawDataset {
    #[allow(dead_code)]
    version: String,
    #[allow(dead_code)]
    generated_at: String,
    /// region -> instance_type -> price entry
    instances: HashMap<String, HashMap<String, InstancePrice>>,
}

struct ProviderDataset {
    /// region -> instance_type -> on-demand $/hour
    prices: HashMap<String, HashMap<String, f64>>,
    expires_at: Instant,
}

// ---------------------------------------------------------------------------
// Caches
// ---------------------------------------------------------------------------

static DATASET_CACHE: Mutex<Option<HashMap<String, ProviderDataset>>> =
    tokio::sync::Mutex::const_new(None);

fn http_client() -> &'static reqwest::Client {
    use std::sync::OnceLock;
    static CLIENT: OnceLock<reqwest::Client> = OnceLock::new();
    CLIENT.get_or_init(|| {
        reqwest::Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .unwrap_or_else(|e| {
                tracing::error!(error = %e, "pricing http client init failed, using default");
                reqwest::Client::new()
            })
    })
}

fn pricing_base_url() -> String {
    std::env::var("KDASHBOARD_PRICING_URL").unwrap_or_else(|_| PRICING_BASE_URL.to_string())
}

fn disk_cache_path(provider: &str) -> Option<PathBuf> {
    let home = dirs::home_dir()?;
    Some(home.join(DISK_CACHE_DIR).join(format!("{provider}.json")))
}

// ---------------------------------------------------------------------------
// Disk cache helpers
// ---------------------------------------------------------------------------

async fn read_disk_cache(provider: &str) -> Option<(String, bool)> {
    let path = disk_cache_path(provider)?;
    let metadata = tokio::fs::metadata(&path).await.ok()?;
    let modified = metadata.modified().ok()?;
    let age = SystemTime::now()
        .duration_since(modified)
        .unwrap_or(Duration::ZERO);
    let fresh = age < DATASET_TTL;

    let contents = tokio::fs::read_to_string(&path).await.ok()?;
    Some((contents, fresh))
}

async fn write_disk_cache(provider: &str, contents: &str) -> Result<()> {
    let path = disk_cache_path(provider).context("home dir unavailable")?;
    if let Some(dir) = path.parent() {
        tokio::fs::create_dir_all(dir)
            .await
            .with_context(|| format!("create_dir_all {dir:?}"))?;
    }
    tokio::fs::write(&path, contents)
        .await
        .with_context(|| format!("write {path:?}"))?;
    Ok(())
}

// ---------------------------------------------------------------------------
// Network fetch
// ---------------------------------------------------------------------------

async fn fetch_dataset_from_network(provider: &str) -> Result<String> {
    let url = format!("{}/{}.json", pricing_base_url(), provider);
    let resp = http_client()
        .get(&url)
        .send()
        .await
        .with_context(|| format!("GET {url}"))?;
    if !resp.status().is_success() {
        anyhow::bail!("pricing fetch returned {}", resp.status());
    }
    let body = resp.text().await.context("read pricing body")?;
    Ok(body)
}

fn parse_dataset(json: &str) -> Result<ProviderDataset> {
    let raw: RawDataset = serde_json::from_str(json).context("parse pricing dataset")?;
    let mut prices: HashMap<String, HashMap<String, f64>> = HashMap::new();
    for (region, types) in raw.instances {
        let mut region_prices: HashMap<String, f64> = HashMap::new();
        for (instance_type, price) in types {
            if let Some(p) = price.ondemand_usd_hour {
                if p > 0.0 {
                    region_prices.insert(instance_type, p);
                }
            }
        }
        if !region_prices.is_empty() {
            prices.insert(region, region_prices);
        }
    }
    Ok(ProviderDataset {
        prices,
        expires_at: Instant::now() + DATASET_TTL,
    })
}

// ---------------------------------------------------------------------------
// Ensure a dataset is available for a provider.
// Order: memory cache → fresh disk → network → stale disk fallback.
// ---------------------------------------------------------------------------

async fn ensure_dataset_loaded(provider: &str) -> bool {
    {
        let cache = DATASET_CACHE.lock().await;
        if let Some(map) = cache.as_ref() {
            if let Some(ds) = map.get(provider) {
                if ds.expires_at > Instant::now() {
                    return true;
                }
            }
        }
    }

    // Try fresh disk cache
    if let Some((contents, fresh)) = read_disk_cache(provider).await {
        if fresh {
            match parse_dataset(&contents) {
                Ok(ds) => {
                    install_dataset(provider, ds).await;
                    return true;
                }
                Err(e) => {
                    tracing::warn!(provider, error = %e, "disk cache corrupt, will refetch");
                }
            }
        }
    }

    // Network fetch
    match fetch_dataset_from_network(provider).await {
        Ok(body) => match parse_dataset(&body) {
            Ok(ds) => {
                if let Err(e) = write_disk_cache(provider, &body).await {
                    tracing::warn!(provider, error = %e, "disk cache write failed");
                }
                install_dataset(provider, ds).await;
                return true;
            }
            Err(e) => {
                tracing::warn!(provider, error = %e, "pricing dataset parse failed");
            }
        },
        Err(e) => {
            tracing::warn!(provider, error = %e, "pricing dataset fetch failed");
        }
    }

    // Last resort: stale disk cache
    if let Some((contents, _)) = read_disk_cache(provider).await {
        match parse_dataset(&contents) {
            Ok(ds) => {
                tracing::info!(provider, "using stale disk cache (network unreachable)");
                install_dataset(provider, ds).await;
                return true;
            }
            Err(e) => {
                tracing::warn!(provider, error = %e, "stale disk cache also unreadable");
            }
        }
    }

    false
}

async fn install_dataset(provider: &str, ds: ProviderDataset) {
    let mut cache = DATASET_CACHE.lock().await;
    let map = cache.get_or_insert_with(HashMap::new);
    map.insert(provider.to_string(), ds);
}

// ---------------------------------------------------------------------------
// Public API: same signature as before — keys are "provider/region/instance_type".
// ---------------------------------------------------------------------------

pub(super) async fn resolve_pricing(nodes: &[NodeInfo]) -> Option<HashMap<String, f64>> {
    if nodes.is_empty() {
        return None;
    }

    let mut providers: Vec<&str> = nodes
        .iter()
        .map(|n| n.provider.as_str())
        .filter(|p| SUPPORTED_PROVIDERS.contains(p))
        .collect();
    providers.sort_unstable();
    providers.dedup();

    if providers.is_empty() {
        return None;
    }

    for provider in &providers {
        ensure_dataset_loaded(provider).await;
    }

    let cache = DATASET_CACHE.lock().await;
    let map = cache.as_ref()?;

    let mut prices: HashMap<String, f64> = HashMap::new();
    for node in nodes {
        let Some(ds) = map.get(&node.provider) else {
            continue;
        };
        let Some(by_region) = ds.prices.get(&node.region) else {
            continue;
        };
        let Some(price) = by_region.get(&node.instance_type) else {
            continue;
        };
        let key = format!("{}/{}/{}", node.provider, node.region, node.instance_type);
        prices.insert(key, *price);
    }

    if prices.is_empty() {
        tracing::info!(
            providers_loaded = providers.len(),
            "pricing dataset loaded but no nodes matched any region/instance_type"
        );
        None
    } else {
        Some(prices)
    }
}

/// Force-refresh pricing by clearing in-memory caches. Disk cache is left in
/// place — next call will refetch from network and overwrite it.
pub async fn refresh_pricing() -> Result<()> {
    {
        let mut cache = DATASET_CACHE.lock().await;
        *cache = None;
    }
    {
        let mut cache = super::calculations::COST_CACHE.lock().await;
        *cache = None;
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    const SAMPLE_JSON: &str = r#"{
        "version": "2026-04-25",
        "generated_at": "2026-04-25T20:17:18Z",
        "source": "https://example.test",
        "provider": "aws",
        "region_count": 1,
        "instance_count": 2,
        "instances": {
            "us-east-1": {
                "m5.large": { "vcpu": 2, "memory_gb": 8, "ondemand_usd_hour": 0.096, "spot_usd_hour": 0.04 },
                "broken": { "vcpu": 1, "memory_gb": 1, "ondemand_usd_hour": null, "spot_usd_hour": null }
            }
        }
    }"#;

    #[test]
    fn parse_dataset_skips_null_prices() {
        let ds = parse_dataset(SAMPLE_JSON).expect("parses");
        let region = ds.prices.get("us-east-1").expect("region");
        assert_eq!(region.get("m5.large").copied(), Some(0.096));
        assert!(region.get("broken").is_none(), "null prices must be filtered");
    }

    #[test]
    fn parse_dataset_drops_zero_priced_entries() {
        let json = r#"{
            "version": "x", "generated_at": "x", "source": "x", "provider": "aws",
            "region_count": 1, "instance_count": 1,
            "instances": { "r": { "i": { "vcpu": 1, "memory_gb": 1, "ondemand_usd_hour": 0, "spot_usd_hour": null } } }
        }"#;
        let ds = parse_dataset(json).expect("parses");
        assert!(
            ds.prices.get("r").is_none(),
            "regions with zero-priced entries become empty and are dropped"
        );
    }

    #[test]
    fn parse_dataset_rejects_invalid_json() {
        match parse_dataset("{ not json") {
            Ok(_) => panic!("expected parse error"),
            Err(e) => assert!(e.to_string().contains("parse"), "unexpected error: {e}"),
        }
    }

    #[tokio::test]
    async fn resolve_pricing_returns_none_when_unsupported_provider() {
        let nodes = vec![NodeInfo {
            name: "n1".into(),
            instance_type: "x".into(),
            provider: "openstack".into(),
            region: "r".into(),
            cpu_capacity: 1.0,
            memory_capacity_bytes: 1.0,
        }];
        assert!(resolve_pricing(&nodes).await.is_none());
    }

    #[tokio::test]
    async fn resolve_pricing_returns_none_for_empty_input() {
        assert!(resolve_pricing(&[]).await.is_none());
    }

    /// Live network test against the actual GitHub release. Ignored by default —
    /// run with `cargo test --lib cost::pricing -- --ignored --nocapture`.
    #[tokio::test]
    #[ignore]
    async fn resolve_pricing_against_real_release() {
        let nodes = vec![
            NodeInfo {
                name: "aws-1".into(),
                instance_type: "m5.large".into(),
                provider: "aws".into(),
                region: "us-east-1".into(),
                cpu_capacity: 2.0,
                memory_capacity_bytes: 8.0 * 1024.0 * 1024.0 * 1024.0,
            },
            NodeInfo {
                name: "gcp-1".into(),
                instance_type: "e2-standard-4".into(),
                provider: "gcp".into(),
                region: "us-central1".into(),
                cpu_capacity: 4.0,
                memory_capacity_bytes: 16.0 * 1024.0 * 1024.0 * 1024.0,
            },
        ];

        let prices = resolve_pricing(&nodes).await.expect("got prices");
        assert!(
            prices.contains_key("aws/us-east-1/m5.large"),
            "missing aws m5.large: {prices:?}"
        );
        assert!(
            prices.contains_key("gcp/us-central1/e2-standard-4"),
            "missing gcp e2-standard-4: {prices:?}"
        );
        for (k, v) in &prices {
            assert!(*v > 0.0 && *v < 100.0, "implausible price for {k}: {v}");
        }
    }
}

File: electron/k8s/pricing.ts
// Cloud pricing for the cluster's nodes: instance types from node labels,
// per-provider datasets (GitHub release assets, conditional GET, 24h TTL,
// disk cache under userData) and the $/core/hour + $/GB/hour rates the cost
// and rightsizing views quote. Extracted from handlers/cost.ts, which mixed
// this with the cost overview itself.

import { promises as fs } from 'node:fs';
import * as nodePath from 'node:path';
import * as os from 'node:os';
import type { V1Node } from '@kubernetes/client-node';

import { atomicWrite } from '../util/fs-atomic.js';
import { getCoreV1Api, onConfigChange } from './client.js';
import { parseCpu, parseMemory } from './quantity.js';

export interface NodeInfo {
  name: string;
  instance_type: string;
  provider: string;
  region: string;
  cpu_capacity: number;
  memory_capacity_bytes: number;
}

export const HOURS_PER_MONTH = 730.0;
export const FALLBACK_CPU_RATE = 0.0325; // $/core/hr
export const FALLBACK_MEM_RATE = 0.0044; // $/GB/hr

// ---------------------------------------------------------------------------
// Node info retrieval
// ---------------------------------------------------------------------------

function detectProvider(labels: Record<string, string>, instanceType: string): string {
  if (
    'eks.amazonaws.com/nodegroup' in labels ||
    'alpha.eksctl.io/nodegroup-name' in labels ||
    instanceType.includes('.')
  ) {
    return 'aws';
  }
  if (
    'cloud.google.com/gke-nodepool' in labels ||
    instanceType.startsWith('e2-') ||
    instanceType.startsWith('n1-') ||
    instanceType.startsWith('n2-') ||
    instanceType.startsWith('n2d-') ||
    instanceType.startsWith('c2-') ||
    instanceType.startsWith('c2d-') ||
    instanceType.startsWith('c3-') ||
    instanceType.startsWith('t2d-')
  ) {
    return 'gcp';
  }
  if ('kubernetes.azure.com/cluster' in labels || instanceType.startsWith('Standard_')) {
    return 'azure';
  }
  return 'unknown';
}

const NODE_INFO_TTL_MS = 30_000; // 30 seconds

/**
 * Memoized node listing shared by get_cost_overview / get_node_costs /
 * get_node_metrics — each used to issue its own full `listNode`. The promise is
 * cached (concurrent callers coalesce into one request); failures are evicted
 * immediately so the next call retries. Invalidated on context switch.
 */
let nodeInfoCache: { promise: Promise<NodeInfo[]>; expiresAt: number } | null = null;

export function getNodeInfo(): Promise<NodeInfo[]> {
  if (nodeInfoCache !== null && nodeInfoCache.expiresAt > Date.now()) {
    return nodeInfoCache.promise;
  }
  const promise = fetchNodeInfo();
  const entry = { promise, expiresAt: Date.now() + NODE_INFO_TTL_MS };
  nodeInfoCache = entry;
  promise.catch(() => {
    // Don't cache failures.
    if (nodeInfoCache === entry) nodeInfoCache = null;
  });
  return promise;
}

async function fetchNodeInfo(): Promise<NodeInfo[]> {
  const core = getCoreV1Api();
  const list = await core.listNode();

  const result: NodeInfo[] = [];
  for (const node of list.items as V1Node[]) {
    const name = node.metadata?.name ?? '';
    const labels = node.metadata?.labels ?? {};

    const instanceType =
      labels['node.kubernetes.io/instance-type'] ??
      labels['beta.kubernetes.io/instance-type'] ??
      '';

    const region =
      labels['topology.kubernetes.io/region'] ??
      labels['failure-domain.beta.kubernetes.io/region'] ??
      '';

    const provider = detectProvider(labels, instanceType);

    const capacity = node.status?.capacity;
    const cpuRaw = capacity?.['cpu'];
    const memRaw = capacity?.['memory'];
    const cpuCap = cpuRaw ? parseCpu(cpuRaw) : 0;
    const memCap = memRaw ? parseMemory(memRaw) : 0;

    // Every node is listed — get_node_metrics needs the capacity of nodes
    // without a cloud instance-type label too (kind, bare metal, k3s), or their
    // CPU/memory meters read 0%. Pricing lookups skip nodes with no
    // instance_type (see resolvePricing / resolveNodeRates).
    result.push({
      name,
      instance_type: instanceType,
      provider,
      region,
      cpu_capacity: cpuCap,
      memory_capacity_bytes: memCap,
    });
  }
  return result;
}

/**
 * Derive per-core and per-GB rates from cloud pricing. Splits the node price
 * 60/40 across CPU and memory.
 */
export function resolveNodeRates(
  nodes: NodeInfo[],
  pricing: Map<string, number>,
): [number, number] {
  for (const node of nodes) {
    const key = `${node.provider}/${node.region}/${node.instance_type}`;
    const pricePerHour = pricing.get(key);
    if (pricePerHour !== undefined) {
      if (node.cpu_capacity > 0 && node.memory_capacity_bytes > 0) {
        const memGb = node.memory_capacity_bytes / (1024.0 * 1024.0 * 1024.0);
        const cpuRate = (pricePerHour * 0.6) / node.cpu_capacity;
        const memRate = (pricePerHour * 0.4) / memGb;
        return [cpuRate, memRate];
      }
    }
  }
  return [FALLBACK_CPU_RATE, FALLBACK_MEM_RATE];
}

onConfigChange(() => {
  nodeInfoCache = null;
});

// ---------------------------------------------------------------------------
// Pricing resolver
// ---------------------------------------------------------------------------

const PRICING_BASE_URL =
  'https://github.com/folio-pro/kdashboard/releases/download/pricing-data';
const SUPPORTED_PROVIDERS = ['aws', 'azure', 'gcp'];
export const DATASET_TTL_MS = 86_400_000; // 24h
const STALE_RETRY_TTL_MS = 300_000; // 5 minutes
const DISK_CACHE_SUBDIR = 'pricing';
const FETCH_TIMEOUT_MS = 30_000;

interface InstancePrice {
  ondemand_usd_hour?: number | null;
}

interface RawDataset {
  instances: Record<string, Record<string, InstancePrice>>;
}

interface ProviderDataset {
  /** region -> instance_type -> on-demand $/hour */
  prices: Map<string, Map<string, number>>;
  expiresAt: number; // epoch ms
}

interface DatasetMeta {
  etag: string | null;
  /** epoch seconds */
  fetched_at: number;
}

/** In-memory pricing cache: provider -> dataset. Null = never populated. */
let datasetCache: Map<string, ProviderDataset> | null = null;

function pricingBaseUrl(): string {
  return process.env['KDASHBOARD_PRICING_URL'] ?? PRICING_BASE_URL;
}

/**
 * Cache under Electron's userData. Resolved lazily and through a dynamic
 * import: this module is reached from handlers the integration suite loads
 * under plain Node, where a static `import { app } from 'electron'` fails at
 * link time. Outside Electron the cache lands under the home directory.
 */
let cacheRootCached: string | null = null;
async function cacheRoot(): Promise<string> {
  if (cacheRootCached) return cacheRootCached;
  let base = os.homedir();
  try {
    const electron = (await import('electron')) as { app?: { getPath(name: string): string } };
    if (electron.app) base = electron.app.getPath('userData');
  } catch {
    // plain Node — keep the home directory
  }
  cacheRootCached = nodePath.join(base, DISK_CACHE_SUBDIR);
  return cacheRootCached;
}

async function datasetPath(provider: string): Promise<string> {
  return nodePath.join(await cacheRoot(), `${provider}.json`);
}

async function metaPath(provider: string): Promise<string> {
  return nodePath.join(await cacheRoot(), `${provider}.meta.json`);
}

function nowEpochSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

function metaIsFresh(meta: DatasetMeta): boolean {
  // A future fetched_at clamps the age to 0, i.e. it counts as fresh.
  const age = Math.max(0, nowEpochSeconds() - meta.fetched_at);
  return age < DATASET_TTL_MS / 1000;
}

const DEFAULT_META: DatasetMeta = { etag: null, fetched_at: 0 };

// --- Disk I/O ---

async function readDiskMeta(provider: string): Promise<DatasetMeta | null> {
  try {
    const raw = await fs.readFile(await metaPath(provider), 'utf8');
    const parsed = JSON.parse(raw) as Partial<DatasetMeta>;
    return {
      etag: typeof parsed.etag === 'string' ? parsed.etag : null,
      fetched_at: typeof parsed.fetched_at === 'number' ? parsed.fetched_at : 0,
    };
  } catch {
    return null;
  }
}

async function writeDiskMeta(provider: string, meta: DatasetMeta): Promise<void> {
  await atomicWrite(await metaPath(provider), JSON.stringify(meta));
}

async function readDiskBody(provider: string): Promise<string | null> {
  try {
    return await fs.readFile(await datasetPath(provider), 'utf8');
  } catch {
    return null;
  }
}

async function writeDiskBody(provider: string, body: string): Promise<void> {
  await atomicWrite(await datasetPath(provider), body);
}

interface FetchedDataset {
  body: string;
  etag: string | null;
}

async function fetchDatasetConditional(
  provider: string,
  prevEtag: string | null,
): Promise<FetchedDataset | null /* null = 304 not modified */> {
  const url = `${pricingBaseUrl()}/${provider}.json`;
  const headers: Record<string, string> = {};
  if (prevEtag) headers['If-None-Match'] = prevEtag;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), FETCH_TIMEOUT_MS);
  let resp: Response;
  try {
    resp = await fetch(url, { headers, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }

  if (resp.status === 304) return null;
  if (!resp.ok) {
    throw new Error(`pricing fetch returned ${resp.status}`);
  }

  const etag = resp.headers.get('etag');
  const body = await resp.text();
  return { body, etag };
}

// --- Parsing ---

function parseDataset(json: string): ProviderDataset {
  const raw = JSON.parse(json) as RawDataset;
  const prices = new Map<string, Map<string, number>>();
  for (const [region, types] of Object.entries(raw.instances ?? {})) {
    const regionPrices = new Map<string, number>();
    for (const [instanceType, price] of Object.entries(types)) {
      const p = price.ondemand_usd_hour;
      if (typeof p === 'number' && p > 0) {
        regionPrices.set(instanceType, p);
      }
    }
    if (regionPrices.size > 0) {
      prices.set(region, regionPrices);
    }
  }
  return { prices, expiresAt: Date.now() + DATASET_TTL_MS };
}

// --- Memory cache helpers ---

function installDataset(provider: string, ds: ProviderDataset): void {
  if (datasetCache === null) datasetCache = new Map();
  datasetCache.set(provider, ds);
}

function memoryHasFresh(provider: string): boolean {
  const ds = datasetCache?.get(provider);
  return ds !== undefined && ds.expiresAt > Date.now();
}

// --- ensure_dataset_loaded main path ---

export async function ensureDatasetLoaded(provider: string): Promise<boolean> {
  if (memoryHasFresh(provider)) return true;

  const meta = (await readDiskMeta(provider)) ?? DEFAULT_META;

  // Fresh disk cache → load directly, no network.
  if (metaIsFresh(meta)) {
    const body = await readDiskBody(provider);
    if (body !== null) {
      try {
        installDataset(provider, parseDataset(body));
        return true;
      } catch {
        // fresh disk cache corrupt — fall through to refetch
      }
    }
  }

  // Stale or missing disk → conditional fetch.
  try {
    const fetched = await fetchDatasetConditional(provider, meta.etag);
    if (fetched !== null) {
      // 200 — body changed (or first fetch). Persist new body + meta.
      await writeDiskBody(provider, fetched.body).catch(() => undefined);
      const newMeta: DatasetMeta = { etag: fetched.etag, fetched_at: nowEpochSeconds() };
      await writeDiskMeta(provider, newMeta).catch(() => undefined);
      try {
        installDataset(provider, parseDataset(fetched.body));
        return true;
      } catch {
        // downloaded dataset failed to parse — fall through
      }
    } else {
      // 304 — body unchanged. Touch meta so we don't revalidate for 24h.
      const touched: DatasetMeta = { etag: meta.etag, fetched_at: nowEpochSeconds() };
      await writeDiskMeta(provider, touched).catch(() => undefined);
      const body = await readDiskBody(provider);
      if (body !== null) {
        try {
          installDataset(provider, parseDataset(body));
          return true;
        } catch {
          // disk body unreadable after 304 — fall through
        }
      }
    }
  } catch {
    // network failed — fall through to stale disk fallback
  }

  // Last resort: stale disk body with a SHORT TTL so the next call retries.
  const body = await readDiskBody(provider);
  if (body !== null) {
    try {
      const ds = parseDataset(body);
      ds.expiresAt = Date.now() + STALE_RETRY_TTL_MS;
      installDataset(provider, ds);
      return true;
    } catch {
      // unparseable — give up
    }
  }
  return false;
}

/**
 * Resolve `(provider, region, instance_type) -> $/hour` for the requested
 * nodes. Returns null when nothing usable could be loaded (caller falls back to
 * hardcoded rates).
 */
export async function resolvePricing(nodes: NodeInfo[]): Promise<Map<string, number> | null> {
  if (nodes.length === 0) return null;

  const providers = Array.from(
    new Set(nodes.map((n) => n.provider).filter((p) => SUPPORTED_PROVIDERS.includes(p))),
  ).sort();

  if (providers.length === 0) return null;

  const loaded = await Promise.all(providers.map((p) => ensureDatasetLoaded(p)));
  const loadedOk = new Set(providers.filter((_, i) => loaded[i]));
  if (loadedOk.size === 0) return null;

  const cache = datasetCache;
  if (cache === null) return null;

  const now = Date.now();
  const prices = new Map<string, number>();
  for (const node of nodes) {
    if (!loadedOk.has(node.provider)) continue;
    const ds = cache.get(node.provider);
    if (ds === undefined || ds.expiresAt <= now) continue;
    const byRegion = ds.prices.get(node.region);
    if (byRegion === undefined) continue;
    const price = byRegion.get(node.instance_type);
    if (price === undefined) continue;
    prices.set(`${node.provider}/${node.region}/${node.instance_type}`, price);
  }

  return prices.size > 0 ? prices : null;
}


/** Drop the in-memory pricing cache; disk state stays (the next load is a cheap 304). */
export function clearPricingCache(): void {
  datasetCache = null;
}

/** The providers whose datasets are currently loaded. */
export function loadedProviders(): string[] {
  return datasetCache ? [...datasetCache.keys()] : [];
}

/**
 * The $/core/hour and $/GB/hour the cost view prices with — cloud pricing
 * when the node instance types resolve, the fallback rates otherwise. Shared
 * with rightsizing so both views quote the same money.
 */
export async function currentRates(): Promise<{ cpu: number; memory: number; source: 'cloud-pricing' | 'fallback' }> {
  let nodes: NodeInfo[] = [];
  try {
    nodes = await getNodeInfo();
  } catch {
    nodes = [];
  }
  const pricing = await resolvePricing(nodes);
  if (pricing !== null && pricing.size > 0 && nodes.length > 0) {
    const [cpu, memory] = resolveNodeRates(nodes, pricing);
    if (Math.abs(cpu - FALLBACK_CPU_RATE) > 0.0001) return { cpu, memory, source: 'cloud-pricing' };
  }
  return { cpu: FALLBACK_CPU_RATE, memory: FALLBACK_MEM_RATE, source: 'fallback' };
}

/**
 * Background pricing revalidation. Every 24h, conditionally re-fetch each
 * ALREADY-CACHED provider's dataset (cheap 304 when unchanged) so pricing never
 * goes stale. Providers that were never loaded are skipped, so a GCP-only
 * cluster never fetches aws/azure. The first tick is at +24h; startup itself
 * loads lazily on the first cost request. Returns a stop function.
 */
export function startPeriodicRefresh(): () => void {
  const timer = setInterval(() => {
    void (async () => {
      for (const provider of loadedProviders()) {
        try {
          await ensureDatasetLoaded(provider);
        } catch {
          // best-effort: a failed revalidation keeps the existing cached data
        }
      }
    })();
  }, DATASET_TTL_MS);
  // Don't keep the process alive solely for this timer.
  timer.unref?.();
  return () => clearInterval(timer);
}
